tests/test_site_apportionment/test_site_apportionment.py

```diff
@@ -634,3 +643,112 @@ def test_sort_rows_missing_cols(self, caplog):
+        [3, 0, "0001", np.nan, np.nan, 1.0, "Check needed", "TMI", "NP30 7ZZ", "202102"], # NaN 601 - Ensure dropped
+    ]
+    input_df = pandasDF(data=input_data, columns=input_cols)
+
+    return input_df
+
+
+class TestCreateSitesDf(object):
+    """Tests for create_sites_df."""
+    
+    @pytest.fixture(autouse=True)
+    def set_attrs(self):
+        """Set class attributes that are passed as parameters."""
+        self.groupby_cols = ["reference", "period"]
+        self.site_cols = ["instance", "601", "602", "postcodes_harmonised"]
+    
+    def test_create_sites_df_raises(self, sites_df_input):
+        """Tests for create_sites_df when invalid data is passed."""
+        # test when a required column is missing
+        no_period = sites_df_input.copy()
+        no_period.drop("period", axis=1, inplace=True)
+        with pytest.raises(KeyError, match=r".*period.*not in index"):
+            create_sites_df(no_period, self.groupby_cols, self.site_cols)
+        # cols passed not in list[str] format
+        with pytest.raises(TypeError, match="unsupported operand type.*"):
+            create_sites_df(sites_df_input, 3, self.site_cols)
+
+    def test_create_sites_df_on_pass(self, sites_df_input):
+        """General tests for create_sites_df."""
+        output = create_sites_df(
+            sites_df_input, 
+            self.groupby_cols,
+            self.site_cols)
+        # assert the resultant dataframe is as expected
+        assert len(output) == 3, f"Output df has {len(output)} row. Expected 3"
+        expected_columns = ["reference",
+                            "period",
+                            "instance",
+                            "601",
+                            "602",
+                            "postcodes_harmonised"]
+        assert sorted(expected_columns) == sorted(output.columns.values)
+        assert np.array_equal(output["602"], [225.0, 100.0, 50.0]), (
+            "Column 602 (percent_col) has unexpected values."
+        )
+        assert np.array_equal(output["instance"], [0, 2, 0]), (
+            "Column 602 (percent_col) has unexpected values."
+        )
+        expected_postcodes = ["NP44 2NZ", "RH12 1XL", "RH12 1XZ"]
+        assert np.array_equal(
+            sorted(output["601"].unique()),
+            sorted(expected_postcodes)
+            ), "Postcodes not as expected"
+
+
+class TestCountDuplicateSites(object):
+    """Tests for count_duplicate_sites."""
+
+    def test_count_duplicate_sites_raises(self, sites_df_input):
+        """Tests for raises from bad data passed to count_duplicate_sites."""
+        # passing dataframe with missing columns
+        with pytest.raises(KeyError, match=".*reference.* not in index"):
+            count_duplicate_sites(sites_df_input.drop("reference", axis=1))
+        # passing wrong type (not a df)
+        with pytest.raises(TypeError, match=".*int.* object is not subscriptable"):
+            count_duplicate_sites(0)
```

Comment:
I like the way you've done these tests for errors. We can use this pattern again.